File: cpp/src/euler053.cpp
// https://projecteuler.net/problem=53

// Combinatoric Selections

// There are exactly ten ways of selecting three from five, 12345:
// 123, 124, 125, 134, 135, 145, 234, 235, 245, and 345

// In combinatorics, we use the notation, \binom 5 3 = 10.
// In general, \binom n r = \dfrac{n!}{r!(n-r)!}, where r ≤ n, n! = n × (n-1) × ... × 3 × 2 × 1, and 0! = 1.

// It is not until n = 23, that a value exceeds one-million: \binom {23} {10} = 1144066.

// How many, not necessarily distinct, values of \binom n r for 1 ≤ n ≤ 100, are greater than one-million?

// Answer: 4075

#include <iostream>

int combinatoric_selections() {
    int count = 0;
    for (int n = 1; n <= 100; ++n) {
        for (int r = 0; r <= n; ++r) {
            long double c = 1.0L;
            for (int i = 1; i <= r; ++i) {
                c *= (n - r + i);
                c /= i;
            }
            if (c > 1000000.0L) ++count;
        }
    }
    return count;
}

#if ! defined UNITTEST_MODE
int main(int argc, char const *argv[]) {
    std::cout << combinatoric_selections() << std::endl;
}
#endif // UNITTEST_MODE

File: go/euler053.go
// https://projecteuler.net/problem=53
//
// Combinatoric Selections
//
// There are exactly ten ways of selecting three from five, 12345: 123, 124, 125, 134, 135, 145, 234, 235, 245, and 345
// In combinatorics, we use the notation, \binom 5 3 = 10.
// In general, \binom n r = \dfrac{n!}{r!(n-r)!}, where r \le n, n! = n \times (n-1) \times ... \times 3 \times 2 \times 1, and 0! = 1.
// It is not until n = 23, that a value exceeds one-million: \binom {23} {10} = 1144066.
// How many, not necessarily distinct, values of \binom n r for 1 \le n \le 100, are greater than one-million?
//
// Answer: 4075

package main

import (
	"fmt"
	"math/big"
)

func binomial(n, k int) *big.Int {
	if k > n-k {
		k = n - k
	}
	res := big.NewInt(1)
	for i := 0; i < k; i++ {
		res.Mul(res, big.NewInt(int64(n-i)))
		res.Div(res, big.NewInt(int64(i+1)))
	}
	return res
}

func main() {
	count := 0
	limit := big.NewInt(1000000)
	for n := 1; n <= 100; n++ {
		for r := 0; r <= n; r++ {
			if binomial(n, r).Cmp(limit) > 0 {
				count++
			}
		}
	}
	fmt.Println(count)
}